RLN: Update API

s1fr0 opened the pull request:
This PR expands public zerokit RLN APIs by adding:

- `set_next_leaf` : set leaf to the next available free index;
- `delete_leaf`: to delete a leaf at a certain index;
- `generate_rln_proof` and `verify_rln_proof`: to verify zk proofs with input/output similar to [Kilic's RLN implementation](https://github.com/kilic/rln);
- `key_gen`: to generate a pair of `(id, commitment)`, with `commitment = PoseidonHash(id)`;
- `hash`: for hashing arbitrary signals to field elements.

Tests for each new rust and C API are also provided.

In order to implement `set_next_leaf`, the module now integrates a slightly customized version of [semaphore-rs](https://github.com/worldcoin/semaphore-rs/) [merkle tree](https://github.com/worldcoin/semaphore-rs/blob/d462a4372f1fd9c27610f2acfe4841fab1d396aa/src/merkle_tree.rs) and [poseidon_tree](https://github.com/worldcoin/semaphore-rs/blob/d462a4372f1fd9c27610f2acfe4841fab1d396aa/src/poseidon_tree.rs) implementations.

This PR will ease integration of zerokit RLN in [nwaku](https://github.com/status-im/nwaku) implementation, with minimal edits to the latter.

oskarth commented on rln/src/ffi.rs:
```diff
@@ -212,7 +332,38 @@ mod test {
+
+        // We now delete all leaves set and check if the root corresponds to the empty tree root
+        // delete calls over indexes higher than no_of_leaves are ignored and will not increase self.tree.next_index
+        for i in 0..2 * no_of_leaves {
```
This line is a bit confusing. Can we pull out the range into a separate variable perhaps? Not do calculation (* no_of_leaves) in range itself

s1fr0 replied:
Sure! Addressed in https://github.com/vacp2p/zerokit/pull/22/commits/28b96a60a631beaebf8386ac5f727322c759d440

oskarth commented on rln/src/merkle_tree.rs:
```diff
@@ -0,0 +1,377 @@
+//! # To do
+//!
+//! * Disk based storage backend (using mmaped files should be easy)
+
```
If I understand it correctly this file is basically same with minor addition to keep track of capacity. Perhaps this can be added as a comment here? So easier to keep up to date.

s1fr0 replied:
Addressed in https://github.com/vacp2p/zerokit/pull/22/commits/28b96a60a631beaebf8386ac5f727322c759d440

oskarth commented on rln/src/poseidon_tree.rs:
```diff
@@ -0,0 +1,23 @@
+// Implementation from https://github.com/worldcoin/semaphore-rs/blob/d462a4372f1fd9c27610f2acfe4841fab1d396aa/src/poseidon_tree.rs
```
What's different with this vs existing impl? Ditto comment above

oskarth reviewed (approved):
Great work! Couldn't find anything immediately wrong, left some comments. Proof is in the pudding though, so let's see how integration works :) Exciting that we are close now!

s1fr0 commented:
Other updates:
- Added circuit for Merkle tree height = 20 (default for `nwaku`)
- Added support for input folder path where circuit and keys can be found (to allow `nwaku` to load circuit resources from custom folder)

s1fr0 commented:
Sorry @staheri14, I didn't see you were added recently as reviewer. Feel free to have a look to the PR and leave your comments: we can revert the merge, if necessary, or address your comments with followup PRs.
